src/cluster_legacy.c

```diff
@@ -3989,28 +4146,64 @@ int clusterSendModuleMessageToTarget(const char *target,
  * Otherwise:
  * Publish this message across the slot (primary/replica).
  * -------------------------------------------------------------------------- */
-void clusterPropagatePublish(robj *channel, robj *message, int sharded) {
-    clusterMsgSendBlock *msgblock;
-
+void clusterPropagatePublish(robj *channel, robj **messages, int count, int sharded) {
+    clusterMsgSendBlock *msgblock, *msgblock_light;
+    int i;
+    msgblock_light = clusterCreatePublishLightMsgBlock(
+        channel, messages, count, sharded ? CLUSTERMSG_TYPE_PUBLISHSHARD_LIGHT : CLUSTERMSG_TYPE_PUBLISH_LIGHT);
     if (!sharded) {
-        msgblock = clusterCreatePublishMsgBlock(channel, message, CLUSTERMSG_TYPE_PUBLISH);
-        clusterBroadcastMessage(msgblock);
-        clusterMsgSendBlockDecrRefCount(msgblock);
+        clusterBroadcastMessage(msgblock_light);
+        clusterMsgSendBlockDecrRefCount(msgblock_light);
+        if (!server.cluster->is_light_hdr_supported) {
+            for (i = 0; i < count; i++) {
+                msgblock = clusterCreatePublishMsgBlock(channel, messages[i], CLUSTERMSG_TYPE_PUBLISH);
+                clusterBroadcastMessage(msgblock);
+                clusterMsgSendBlockDecrRefCount(msgblock);
+            }
+        }
```

Comment:
As mentioned in another comment, I think we should ignore `clusterBroadcastMessage` here and just loop over the nodes. We can keep the clusterBroadcastMessage function simple and I think it will be simpler even here. We already do a loop over nodes for sharded pubsub so we can do the same for unsharded and sharded, just loop over another list of nodes.

I wanted to see how complex it gets, so I wrote the (untested!) code below. With this idea, we don't need to store `server.cluster->is_light_hdr_supported` at all. It looks simpler to me, but I probably missed something...

(...) I just realized that the shard nodes is a list, but all nodes is a dict, so the iteration is different. I did a quickfix to the iteration stuff, which looks a bit ugly. :) It avoids duplicated code though. WDYT?

```C
void clusterPropagatePublish(robj *channel, robj **messages, int count, int sharded) {
    /* Lazy create msg blocks later when we need them. */
    clusterMsgSendBlock *msgblock_light = NULL;     /* Just one, multi-message */
    clusterMsgSendBlock *msgblocks[count] = {NULL}; /* Heavy, one per message */
    dictIterator di;                                /* For unsharded */
    listIter li;                                    /* For sharded */
    if (sharded) {
        list *nodes_for_slot = clusterGetNodesInMyShard(server.cluster->myself);
        serverAssert(nodes_for_slot != NULL);
        listRewind(nodes_for_slot, &li);
    } else {
        dictInitSafeIterator(&di, server.cluster->nodes);
    }
    do {
        clusterNode *node;
        if (sharded) {
            listNode *ln = listNext(&li);
            if (!ln) break;
            node = listNodeValue(ln);
        } else {
            dictEntry *de = dictNext(&di);
            if (!de) break;
            node = dictGetVal(de);
        }
        if (node->flags & (CLUSTER_NODE_MYSELF | CLUSTER_NODE_HANDSHAKE)) continue;
        if (nodeSupportsLightMsgHdr(node)) {
            if (msgblock_light == NULL) {
                /* Lazy creation */
                msgblock_light = clusterCreatePublishLightMsgBlock(...);
            }
            clusterSendMessage(node->link, msgblock_light);
        } else {
            /* Heavy msgblock, one for each pubsub message. */
            for (i = 0; i < count; i++) {
                if (msgblocks[i] == NULL) {
                    /* Lazy create msgblock */
                    msgblocks[i] = clusterCreatePublishMsgBlock(...);
                }
                clusterSendMessage(node->link, msgblocks[i]);
            }
        }
    }

    /* Cleanup what we created. */
    if (!sharded) dictResetIterator(&di);
    if (light_msgblock != NULL) clusterMsgSendBlockDecrRefCount(light_msgblock);
    if (msgblocks[0] != NULL) {
        for (i = 0; i < count; i++) {
            clusterMsgSendBlockDecrRefCount(msgblocks[i]);
        }
    }
}
```

Reply:
I actually like this, no problems of maintaining the state and no changes to broadcast, I will see if I can prettify this

Reply:
We could create a special wrapper-iterator that can be initialized with either a dict or a list.

Reply:
That would create a lot of typecasting and less readable

Reply:
I wanted to see for myself so I wrote the untested pseudo-code below. :) There is a union involved, no type casting. clusterPropagatePublish is more readable but the iterator has some code and a struct. It's more code in total but each function is perhaps more readable?

If we need it in more places, perhaps it would be worth it, but now it's just too much code. :)

Usage in clusterPropagatePublish():

```C
void clusterPropagatePublish(robj *channel, robj **messages, int count, int sharded) {
    ...
    DictOrListIterator iter;
    if (sharded) {
        list *nodes_for_slot = clusterGetNodesInMyShard(server.cluster->myself);
        serverAssert(nodes_for_slot != NULL);
        iterInitForList(&iter, nodes_for_slot);
    } else {
        iterInitForDict(&iter, server.cluster->nodes);
    }
    clusterNode *node;
    while ((node = iterNext(&iter)) != NULL) {
        /* Send message to this node */
        ...
    }
    iterReset(&iter);
```

Implementation of dict-or-list-iterator:

```C
typedef struct {
    enum { ITER_DICT, ITER_LIST } type;
    union {
        dictIterator di;
        listIter li;
    };
} DictOrListIterator;

void iterInitForDict(DictOrListIterator *iter, dict *d) {
    iter->type = ITER_DICT;
    dictInitSafeIterator(&iter->di, d);
}

void iterInitForList(DictOrListIterator *iter, list *l) {
    iter->type = ITER_LIST;
    listRewind(l, &iter->li);
}

void *iterNext(DictOrListIterator *iter) {
    switch (iter->type) {
    case ITER_DICT:
        dictEntry *de = dictNext(&iter->di);
        return de ? dictGetVal(de) : NULL;
    case ITER_LIST:
        listNode *ln = listNext(&iter->li);
        return ln ? listNodeValue(ln) : NULL;
    default:
        serverPanic("bad type");
    }
}

void iterReset(DictOrListIterator *iter) {
    if (iter->type == ITER_DICT) dictResetIterator(&iter->di);
}
```

Reply:
I did add the new Iterator and make the broadcast code simpler to read as well as added a new shardbroadcast method.